Revive of PR #1686.

yoshisatoyanagisawa opened the pull request:
I used to work on the specification update to support the ServiceWorker static routing API
(https://github.com/WICG/service-worker-static-routing-api) https://github.com/w3c/ServiceWorker/pull/1686

However, I accidentally closed it by force-sync to the ServiceWorker specification's repository HEAD.

This CL is for reviving it.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/1.html" title="Last updated on Dec 6, 2023, 9:42 AM UTC (acdd79d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/1/4711a07...acdd79d.html" title="Last updated on Dec 6, 2023, 9:42 AM UTC (acdd79d)">Diff</a>

yoshisatoyanagisawa committed:
Revive of PR #1686.

I used to work on the specification update to support the
ServiceWorker static routing API
(https://github.com/WICG/service-worker-static-routing-api)
https://github.com/w3c/ServiceWorker/pull/1686

However, I accidentally closed it by force-sync to the ServiceWorker
specification's repository HEAD.

This CL is for reviving it.

yoshisatoyanagisawa commented:
@domenic @sisidovski Can I ask your review for this?
My preference to this PR is dropping something controversial and move forward with something we can agree.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3235,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+    : Output
+    :: a boolean
+
+    1. If running the [=VerifyURLPatternComponent=] algorithm with |pattern|'s [=urlpattern/protocol component=] returns false, return false.
```
Let's wait until https://github.com/whatwg/urlpattern/pull/199 lands, and then we can get rid of this whole algorithm (and its sub-algorithm). Instead we can just say `|pattern| [=URLPattern/has regexp groups=]`.

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3235,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"] is the empty string, return false.
```
Why? That is not how URLPattern works elsewhere on the platform.

yoshisatoyanagisawa replied:
Removed.  That was relevant when I first implemented the spec PR, but I do not think it is true any more.

domenic commented on docs/index.bs:
```diff
@@ -1545,6 +1563,46 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      };
+
+      dictionary RouterCondition {
+        USVString urlPattern;
```
Wait for https://github.com/whatwg/urlpattern/pull/199 to land. Then use `URLPatternCompatible`.

sisidovski replied:
+1, also this update already landed in chromium.

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3235,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"] is the empty string, return false.
+      1. Let |rawPattern| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"].
+      1. If |rawPattern| is a [=string=], then:
```
After https://github.com/whatwg/urlpattern/pull/199 lands you can replace this whole thing with

```html
Let |pattern| be the result of [=building a URLPattern from a Web IDL value=] |rawPattern| given |baseURL| and |serivceWorker|'s [=service worker/global object=]'s [=relevant realm=].
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3235,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+        1. Set |pattern| to the result of constructing a {{URLPattern}} using the {{URLPattern/URLPattern(input)}} constructor steps given |rawPattern|. If those steps throw, catch the exception and return false.
+      1. Otherwise, return false.
+      1. If running the [=VerifyURLPattern=] algorithm with |pattern| returns false, return false.
+      1. Set |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"]'s associated {{URLPattern}} to |pattern|.
```
As discussed in the previous PR, you cannot create an "associated URLPattern" concept for a string (or URLPatternCompatible), which is what `|rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"]` is.

You need some sort of parallel structure, representing the fully parsed routing rule. That will probably be a new [struct](https://infra.spec.whatwg.org/#struct) you define with two members: URL pattern (a `URLPattern`), and source (a `RouterSourceEnum`).

This algorithm will then be responsible for parsing the Web IDL dictionary into this struct, so that you can store it in the "static router rules object".

yoshisatoyanagisawa replied:
Instead of associating the URLPattern, I tried to build them on-demand.
Chromium implementation just verifies the URLPattern, and do not reuse the built pattern here.

domenic replied:
I guess that works too; it feels inefficient and a bit silly, but since parsing gives the same result each time (and the base URL cannot change), it is OK.

domenic commented on docs/index.bs:
```diff
@@ -207,6 +223,8 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
     A [=/service worker=] has an associated <dfn>all fetch listeners are empty flag</dfn>. It is initially unset.
 
+    A [=/service worker=] has an associated <dfn>static router rules object</dfn>. It is initially unset.
```
Can you state its type? Per the below, I think it should be a list of router rules. (Which makes the use of the "object" suffix kind of strange.)

yoshisatoyanagisawa replied:
renamed to "list of router rules".

domenic commented on docs/index.bs:
```diff
@@ -1545,6 +1563,46 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+        USVString urlPattern;
+      };
+
+      enum RouterSourceEnum { "network" };
```
```suggestion
      enum RouterSource { "network" };
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -1545,6 +1563,46 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+    <pre class="idl">
+      [Exposed=ServiceWorker]
+      interface InstallEvent : ExtendableEvent {
+        Promise&lt;undefined&gt; registerRouter((RouterRule or sequence&lt;RouterRule&gt;) rules);
```
I thought the name was changed?

sisidovski replied:
Yes. This should be `addRoutes((RouterRule or sequence&lt;RouterRule&gt;) rules);`

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -1545,6 +1563,46 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+        1. Let |routerRules| be a list of {{RouterRule}} dictionaries.
+        1. If |rules| is a {{RouterRule}} dictionary, set |rules| to &#x00AB; |rules| &#x00BB;.
+        1.  for each |rule| in |rules|:
```
```suggestion
        1.  For each |rule| of |rules|:
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -133,6 +133,22 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
         text: obtain a storage key; url: obtain-a-storage-key
         text: storage key/equals; url: storage-key-equals
 
+spec: urlpattern; urlPrefix: https://wicg.github.io/urlpattern/
```
After https://github.com/whatwg/urlpattern/pull/199 is merged and you make the other changes mentioned here, this section should be able to be deleted.

In general, any time something is added to this section it means you are poking into the non-exposed internals of some other spec, which is bad. It is better to export the necessary terms (like I suggest in https://github.com/whatwg/urlpattern/pull/199#pullrequestreview-1754767750), or create abstraction boundaries (like https://github.com/whatwg/urlpattern/pull/199 does with "has regexp groups" and "building a URLPattern from a Web IDL value").

yoshisatoyanagisawa replied:
Without:
```
spec: urlpattern; urlPrefix: https://wicg.github.io/urlpattern/
    type: dfn
        text: match; for: urlpattern; url: match
```
bikeshed seems to show the error like:
```
LINE ~3251: No 'dfn' refs found for 'match' with for='['urlpattern']'.
[=urlpattern/match=]
```
Do you know how I can mitigate that?

So do "building a urlpattern from a webidl value"?

domenic replied:
It looks like these were not exported correctly; I sent https://github.com/whatwg/urlpattern/pull/201 to fix it.

It also takes about a day for the database to update, and to get the changes locally you might need to run `bikeshed update`.

yoshisatoyanagisawa replied:
Unfortunately, I still see the following error if I remove the lines.
```
LINE ~3265: No 'dfn' refs found for 'match' with for='['urlpattern']'.
[=urlpattern/match=]
```

I have executed `bikeshed update` and I do not need to write the entry for "building a urlpattern from a webidl value", though.

If I remove "urlpattern" and just use "match", I see:
```
LINE ~3265: Multiple possible 'match' dfn refs.
Arbitrarily chose https://drafts.css-houdini.org/css-typed-om-1/#cssnumericvalue-match
To auto-select one of the following refs, insert one of these lines into a <pre class=link-defaults> block:
spec:css-typed-om-1; type:dfn; text:match
spec:selectors-4; type:dfn; text:match
spec:urlpattern; type:dfn; text:match
spec:css2; type:dfn; text:match
[=match=]
```
Ah, I should write `spec:urlpattern; type:dfn; text:match` to `<pre class=link-defaults>` block instead.
done.

sisidovski commented on docs/index.bs:
```diff
@@ -3175,6 +3235,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
   </section>
 
+  <section algorithm>
+    <h3 id="verify-router-rule-algorithm"><dfn>VerifyRouterRule</dfn></h3>
```
Let's consider in a follow up PR, but we may be able to cover some points concerned in https://github.com/WICG/service-worker-static-routing-api/issues/14 here. More specifically, throw an error if the router rules contain "fetch-handler" but the SW doesn't have any fetch handers or empty fetch handler flag is true.

yoshisatoyanagisawa replied:
I do not have a plan to implement fetch-event source in this PR yet.  Therefore, it might not be affected by  WICG/service-worker-static-routing-api#14.  Or, is it better to have that in this PR?

sisidovski replied:
No, I mean the follow-up PR is totally fine. I just wanted to put a comment so that we don't forget about it :)

yoshisatoyanagisawa replied:
Sure.  Thanks for the clarification.

sisidovski commented on docs/index.bs:
```diff
@@ -3175,6 +3235,78 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"] is the empty string, return false.
+      1. Let |rawPattern| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"].
+      1. If |rawPattern| is a [=string=], then:
+        1. Let |baseURL| be |serviceWorker|'s [=service worker/script url=].
```
Just to clarify, `[=service worker/script url=]` is a main script URL, correct?

yoshisatoyanagisawa replied:
I understand so.
It refers https://w3c.github.io/ServiceWorker/#dfn-script-url

yoshisatoyanagisawa committed:
Introduce changes upon whatwg/urlpattern#199.

yoshisatoyanagisawa committed:
Fixed the parse error.

yoshisatoyanagisawa committed:
clean up.

yoshisatoyanagisawa committed:
another dfn fix.

yoshisatoyanagisawa committed:
Rename of static router rules object.

yoshisatoyanagisawa committed:
Stop associating the URLPattern and construct URLPattern on-demand.

domenic commented on docs/index.bs:
```diff
@@ -1545,6 +1552,44 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      {{InstallEvent/addRoutes(rules)}} registers this service worker the rules to offload simple tasks that the fetch handler does.
+
+      <dfn method for="InstallEvent"><code>addRoutes(|rules|)</code></dfn> method *must* run these steps:
```
```suggestion
      The <dfn method for="InstallEvent"><code>addRoutes(|rules|)</code></dfn> method steps are:
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3222,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. Let |rawPattern| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"].
```
Since `urlPattern` is not `required` in the Web IDL, this might not exist. You need to handle that case here.

yoshisatoyanagisawa replied:
Good catch!
I have updated it while considering the future update there.
I made non-existing of urlPattern as a verify error.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3222,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      1. Let |rawPattern| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"].
+      1. Let |pattern| be the result of running <a>ParseURLPattern</a> algorithm passing |rawPattern| and |serviceWorker|. If this throws an exception, catch it and return false.
+      1. If |pattern| [=URLPattern/has regexp groups=] returns true, return false.
```
```suggestion
      1. If |pattern| [=URLPattern/has regexp groups=], then return false.
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3222,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: {{RouterSource}} or null
+
+      1. [=list/For each=] |rule| of |serviceWorker|'s [=service worker/list of router rules=]:
+          1. Let |rawPattern| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"].
```
As above, you need to handle the case where this is not present. (Or, make it `required`.)

yoshisatoyanagisawa replied:
I think this is already resolved by "If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"] does not [=map/exist=], continue." written in the line above.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3222,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: {{URLPattern}}
+
+      1. Let |baseURL| be |serviceWorker|'s [=service worker/script url=].
+      1. Return the result of [=urlpattern/building a URLPattern from a WebIDL value=] |rawPattern| given |baseURL| and |serviceWorker|'s [=service worker/global object=]'s [=relevant realm=].
```
```suggestion
      1. Return the result of [=building a URLPattern from a WebIDL value=] |rawPattern| given |baseURL| and |serviceWorker|'s [=service worker/global object=]'s [=relevant realm=].
```

This change should probably allow you to remove line 139.

However, after https://github.com/whatwg/urlpattern/pull/201 is merged, you will need to update accordingly, from `WebIDL` to `Web IDL`.

yoshisatoyanagisawa replied:
thanks.  I have updated this.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3222,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+  <section algorithm>
+    <h3 id="parse-urlpattern-algorithm"><dfn>ParseURLPattern</dfn></h3>
+      : Input
+      :: |rawPattern|, a {{RouterCondition/urlPattern}}"
```
```suggestion
      :: |rawPattern|, a [=string=]
```

(The type is `[=string=]`; `{{RouterCondition/urlPattern}}` is a field name, not a type.)

yoshisatoyanagisawa replied:
done.

sisidovski replied:
I may be misunderstanding, but is the type string here? I thought `URLPatternCompatible` is the correct type.

domenic replied:
Oh, this is a good catch!!

yoshisatoyanagisawa replied:
Thanks for noticing this.  done.

domenic reviewed (commented):
Only one significant issue, about what happens if the `"urlPattern"` dictionary member is not present.

yoshisatoyanagisawa committed:
Addressed some of comments.

yoshisatoyanagisawa committed:
Update the case without the urlPattern condition.

Considering the future update of the condition, we should mark the case
handled as one of conditions.  Also, we need to make it extensible.

yoshisatoyanagisawa committed:
Fix the urlPattern existence handling.

It was flipped by mistake.

yoshisatoyanagisawa committed:
Remove the "spec: urlpattern; ..." line.

yoshisatoyanagisawa committed:
Simplified the if statement on urlPattern condition does not exist.

yoshisatoyanagisawa committed:
Fix typo.

yoshisatoyanagisawa commented:
> Only one significant issue, about what happens if the "urlPattern" dictionary member is not present.

I have updated the specification to make the case handled.
In that case, `VerifyRouterRule()` returns false, and `GetRouterSource()` ignores.

domenic commented on docs/index.bs:
```diff
@@ -1545,6 +1547,44 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+        1.  For each |rule| of |rules|:
+            1. If running [=VerifyRouterRule=] algorithm with |rule| and [=/service worker=] returns false, <a>throw</a> a <code>TypeError</code>.
+            1. Append |rule| to |routerRules|.
+        1. Set [=/service worker=]'s [=service worker/list of router rules=] to |routerRules|.
```
This is not a specific-enough reference to service worker; it's just referring to the service worker type. https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/1.html#wait-until-method step 5.2.1 seems to have a good way of getting to the right object: "current global object's associated service worker". (Note that the "service worker" there links to https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/1.html#serviceworkerglobalscope-service-worker instead of https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/1.html#dfn-service-worker .)

yoshisatoyanagisawa replied:
Clarified which service worker to use.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3217,55 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"] does not [=map/exist=], Return false.
```
```suggestion
      1. If |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"] does not [=map/exist=], return false.
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3217,55 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. Let |rawPattern| be |rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"].
+          1. Let |pattern| be the result of running <a>ParseURLPattern</a> algorithm passing |rawPattern| and |serviceWorker|.
+          1. If running [=match=] with |pattern| and |request|'s [=request/URL=] returns null, [=continue=].
+          1. Return |rule|'s {{RouterRule/source}}.
```
```suggestion
          1. Return |rule|["{{RouterRule/source}}"].
```

domenic commented on docs/index.bs:
```diff
@@ -3067,6 +3107,8 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
       1. If |request|'s [=request/destination=] is either "<code>embed</code>" or "<code>object</code>", then:
           1. Return null.
+      1. Else if |registration|'s <a>active worker</a>'s [=service worker/list of router rules=] is set:
+          1. If running [=GetRouterSource=] algorithm with <a>active worker</a> and |request| returns "network", return null.
```
```suggestion
          1. If running [=GetRouterSource=] with |registration|'s <a>active worker</a> and |request| returns "network", return null.
```

yoshisatoyanagisawa replied:
I guess the goal for this is clarifying |registration|.  I kept algorithm as-is.

domenic commented on docs/index.bs:
```diff
@@ -3175,6 +3217,55 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
   </section>
 
+  <section algorithm>
+    <h3 id="parse-urlpattern-algorithm"><dfn>ParseURLPattern</dfn></h3>
```
I just realized other algorithm names are written as "Parse URL Pattern", instead of "ParseURLPattern" and "VerifyRouterRule". It'd be good to stay consistent and include spaces.

yoshisatoyanagisawa replied:
Thank you for noticing it.  I renamed them.

domenic reviewed (approved):
LGTM with nits

yoshisatoyanagisawa committed:
Merged the simple fixes.

yoshisatoyanagisawa committed:
siple fix.

yoshisatoyanagisawa committed:
Rename algorithm names to align with other algorithm names.

yoshisatoyanagisawa committed:
Make which service worker to use clear.

yoshisatoyanagisawa committed:
Put algorithm after the algorithm.

yoshisatoyanagisawa commented:
@sisidovski Can I ask you to take another look?  Does it look good?

sisidovski commented on docs/index.bs:
```diff
@@ -3175,6 +3217,55 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: {{URLPattern}}
+
+      1. Let |baseURL| be |serviceWorker|'s [=service worker/script url=].
+      1. Return the result of [=building a URLPattern from a WebIDL value=] |rawPattern| given |baseURL| and |serviceWorker|'s [=service worker/global object=]'s [=relevant realm=].
```
`building a URLPattern from a Web IDL value` from https://github.com/whatwg/urlpattern/pull/201.

yoshisatoyanagisawa replied:
done.  good catch!

yoshisatoyanagisawa committed:
Fixed typo.

sisidovski reviewed (approved):
LGTM w/ above comments.

yoshisatoyanagisawa committed:
Fix "Parse URL Pattern" argument.
